Handle Facet endValue; test individual XML elements

khaeru opened the pull request:
This PR:
- Adjusts sdmxml.Reader.parse_facet to handle all attributes of FacetType using a regex.
- Adds a general method for fast, atomic tests of parsing single XML elements.

dr-leo commented on pandasdmx/reader/sdmxml.py:
```diff
@@ -1208,15 +1208,9 @@ def parse_facet(self, elem):
         if isinstance(value_type, str):
             value_type = FacetValueType[value_type[0].lower() + value_type[1:]]
```
This method is not self-explanatory and contains inefficient code. The copy and pop operations are not needed. Please explain why snake-case is to be returned.

khaeru replied:
> The copy and pop operations are not needed.

The remaining elements of `attr` is used in the subsequent loop. The 'textType' XML attribute needs to be processed *first*, in order to create the Facet object, which is next populated by the loop. But `textType` is not guaranteed to be the first XML attribute appearing in `elem.attrib`; e.g. it could be the last. It is much simpler to pop this element from the dict and then loop over the remainder, than to handle it within a loop where it would appear in an unpredictable order.

`copy()` is needed because `elem.attrib.pop()` would destructively modify the XML document being read. This is not the expected behaviour of a "reader."

I will update/expand the comments. Since copy/pop are used consistently throughout the reader in this way, some of this can go in a file-level comment that explains the design.

> why snake-case is to be returned

All of the code follows [PEP 8 naming conventions](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables):

> **Method Names and Instance Variables**
>
> Use the function naming rules: lowercase with words separated by underscores as necessary to improve readability.

i.e., snake case. All of the attributes of the IM classes, including FacetType, follow PEP 8. SDMX-ML uses lower camel case for attribute names.

dr-leo commented:
A more succinct and arguably faster alternative to the
copy...Pop...loop pattern is as follows:

`for k,v in elem.attr.items():
    if k == 'text_type': do_x
    else: do_y`

This comment applies throughout the core.

Btw: I do not see the Benefit of the isinstance check. If we are
concerned that the value is not guaranteed to be str, will the
remaining Code work? If we are sure that is str (save pathological
cases we can't and should not handle), then let's drop the check.

 Finally, I recommend to pre-compile the regex's as class Attributes
rather than compiling them ad hoc upon each method call or even at
each Iteration as it is now.




On 11/02/2020, Paul Natsuo Kishimoto <notifications@github.com> wrote:
> khaeru commented on this pull request.
>
>
>
>> @@ -1208,15 +1208,9 @@ def parse_facet(self, elem):
>          if isinstance(value_type, str):
>              value_type = FacetValueType[value_type[0].lower() +
> value_type[1:]]
>
>> The copy and pop operations are not needed.
>
> The remaining elements of `attr` is used in the subsequent loop. The
> 'textType' XML attribute needs to be processed *first*, in order to create
> the Facet object, which is next populated by the loop. But `textType` is not
> guaranteed to be the first XML attribute appearing in `elem.attrib`; e.g. it
> could be the last. It is much simpler to pop this element from the dict and
> then loop over the remainder, than to handle it within a loop where it would
> appear in an unpredictable order.
>
> `copy()` is needed because `elem.attrib.pop()` would destructively modify
> the XML document being read. This is not the expected behaviour of a
> "reader."
>
> I will update/expand the comments. Since copy/pop are used consistently
> throughout the reader in this way, some of this can go in a file-level
> comment that explains the design.
>
>> why snake-case is to be returned
>
> All of the code follows [PEP 8 naming
> conventions](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables):
>
>> **Method Names and Instance Variables**
>>
>> Use the function naming rules: lowercase with words separated by
>> underscores as necessary to improve readability.
>
> i.e., snake case. All of the attributes of the IM classes, including
> FacetType, follow PEP 8. SDMX-ML uses lower camel case for attribute names.
>
> --
> You are receiving this because your review was requested.
> Reply to this email directly or view it on GitHub:
> https://github.com/dr-leo/pandaSDMX/pull/132#discussion_r377565091

dr-leo commented:
PS: Don't get me wrong. The copy.pop Thing is to some extent a
question of taste.

On 11/02/2020, Leo <fhaxbox66@googlemail.com> wrote:
> A more succinct and arguably faster alternative to the
> copy...Pop...loop pattern is as follows:
>
> `for k,v in elem.attr.items():
>     if k == 'text_type': do_x
>     else: do_y`
>
> This comment applies throughout the core.
>
> Btw: I do not see the Benefit of the isinstance check. If we are
> concerned that the value is not guaranteed to be str, will the
> remaining Code work? If we are sure that is str (save pathological
> cases we can't and should not handle), then let's drop the check.
>
>  Finally, I recommend to pre-compile the regex's as class Attributes
> rather than compiling them ad hoc upon each method call or even at
> each Iteration as it is now.
>
>
>
>
> On 11/02/2020, Paul Natsuo Kishimoto <notifications@github.com> wrote:
>> khaeru commented on this pull request.
>>
>>
>>
>>> @@ -1208,15 +1208,9 @@ def parse_facet(self, elem):
>>          if isinstance(value_type, str):
>>              value_type = FacetValueType[value_type[0].lower() +
>> value_type[1:]]
>>
>>> The copy and pop operations are not needed.
>>
>> The remaining elements of `attr` is used in the subsequent loop. The
>> 'textType' XML attribute needs to be processed *first*, in order to
>> create
>> the Facet object, which is next populated by the loop. But `textType` is
>> not
>> guaranteed to be the first XML attribute appearing in `elem.attrib`; e.g.
>> it
>> could be the last. It is much simpler to pop this element from the dict
>> and
>> then loop over the remainder, than to handle it within a loop where it
>> would
>> appear in an unpredictable order.
>>
>> `copy()` is needed because `elem.attrib.pop()` would destructively modify
>> the XML document being read. This is not the expected behaviour of a
>> "reader."
>>
>> I will update/expand the comments. Since copy/pop are used consistently
>> throughout the reader in this way, some of this can go in a file-level
>> comment that explains the design.
>>
>>> why snake-case is to be returned
>>
>> All of the code follows [PEP 8 naming
>> conventions](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables):
>>
>>> **Method Names and Instance Variables**
>>>
>>> Use the function naming rules: lowercase with words separated by
>>> underscores as necessary to improve readability.
>>
>> i.e., snake case. All of the attributes of the IM classes, including
>> FacetType, follow PEP 8. SDMX-ML uses lower camel case for attribute
>> names.
>>
>> --
>> You are receiving this because your review was requested.
>> Reply to this email directly or view it on GitHub:
>> https://github.com/dr-leo/pandaSDMX/pull/132#discussion_r377565091
>

khaeru commented:
> A more succinct and arguably faster alternative to the copy...Pop...loop pattern is as follows: `for k,v in elem.attr.items(): if k == 'text_type': do_x else: do_y`

```
# similar to Facet
class Foo:
    info = {}

    def __init__(self, x):
        self.x = x

# XML attributes may be in any order
attr = {'otherValue': 1, 'textType': 2, 'differentValue': 3}

# This does not work, because id == 'otherValue' on the first iteration
for id, value in attr.items():
    if id == 'textType':
        # Instantiate the class
        f = Foo(value)
    else:
        # raises NameError on first iteration because f has not been created
        f.info[id] = value
```
The latter can only work if we assume the `attr` are in a specific order—again, we can not assume this.

Besides the pattern I described, a second pattern that does work, and is used in some of the other parse methods, is to consume child elements and/or attributes into a keyword arguments dict, and then instantiate the model class as a late step. See parse_dimension, for example: https://github.com/dr-leo/pandaSDMX/blob/67d818091089bd9d355d15cec4be723be503ae23/pandasdmx/reader/sdmxml.py#L1103-L1125

In any case, notice that this method _still_ uses copy/pop to modify the attributes into the keyword arguments needed to instantiate the class. For child elements, pop is used in order to enable the assertion `len(values) == 0`. These assertions are a key sanity check to ensure that we are not throwing away child elements or attributes containing data without storing them properly.

> let's drop the [isinstance] check
>
> I recommend to pre-compile the regex's

These are helpful suggestions, thank you.

dr-leo commented:
In Addition to my previous comment: Sanity checks are importent. But we should do them only explicitly when pydantic's checks aren't sufficient. E.g., if we loop over FacetTypeVylue Attributes, pydantic will throw an exception if some value is invalid. This is not to say that `assert len(mylist) == 0` is never a good idea. Same holds true for `mylist.pop()`. But as shown above, I think in many situations we are better off without the copy-pop Pattern.

khaeru commented:
@dr-leo per your comment in a380dd1:
```
# The following handles xml nodes tagged 'TextFormat' 
# TODO: Check if EnumerationFormat is properly handled? '. 
```

The METHOD global variable ensures this same method is used for both elements. EnumerationFormat occurs several times in tests/data/exr/1/structure.xml, so it is tested.

Looking at the SDMX-ML standard (specifically the file schemas/SDMXStructureBase.xsd in §3B):
- SDMX-ML defines a 'CodeDataType' that is a subset of 'FacetValueType':
  > CodeDataType is a restriction of the basic data types that are applicable to codes. Although some of the higher level time period formats are perimitted, [sp?] it should be noted that any value which contains time (which includes a time zone offset) is not allowable as a code identifier.
- The former is not mentioned at all in the Information Model.

So this code works for EnumerationFormat, but at the moment it will silently accept values for the textType attribute that are not compatible with the XSD. However, I agree that strict validation of XML according to the XSD is out-of-scope for this PR, so I look forward to you opening that new issue to discuss it. I think that is an enhancement that can wait until after the 1.0 release.

I will make another commit to clean up the comments and then, assuming this addresses your concerns, merge the PR.

khaeru commented:
CI checks pass.

Because #131 disabled running the tests of specific sources for PRs:
- Test coverage appears to decrease; but this is spurious.
- I manually ran the `--remote-data` tests on my machine, and confirm that none fail.
